File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: bool
        """
        if head == None: 
            return True

        prev = None
        slow = fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        curr = slow

        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp

        while prev:
            if head.val != prev.val:
                return False
            head = head.next
            prev = prev.next

        return True

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool isPalindrome(struct ListNode* head) {
    if(head == NULL) return head;

    struct ListNode *slow, *fast, *prev = NULL, *curr;
    slow = fast = head;

    while(fast != NULL && fast -> next != NULL) {
        slow = slow -> next;
        fast = fast -> next -> next;
    }

    curr = slow;

    while(curr != NULL) {
        struct ListNode* temp = curr -> next;
        curr -> next = prev;
        prev = curr;
        curr = temp;
    }

    while(prev != NULL) {
        if(head -> val != prev -> val) return false;
        head = head -> next;
        prev = prev -> next;
    }

    return true;
}